src/registers/debug.rs

```diff
@@ -0,0 +1,599 @@
+            _ => unreachable!(),
+        }
+    }
+
+    /// Returns the global breakpoint enable flag of the provided debug address register.
+    pub fn global_breakpoint_enable(n: DebugAddressRegisterNumber) -> Self {
+        match n.0 {
+            0 => Self::GLOBAL_BREAKPOINT_0_ENABLE,
+            1 => Self::GLOBAL_BREAKPOINT_1_ENABLE,
+            2 => Self::GLOBAL_BREAKPOINT_2_ENABLE,
+            3 => Self::GLOBAL_BREAKPOINT_3_ENABLE,
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// The condition for a hardware breakpoint.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(u8)]
+pub enum HwBreakpointCondition {
+    /// Instruction execution
+    InstructionExecution = 0b00,
+
+    /// Data writes
+    DataWrites = 0b01,
+
+    /// I/O reads or writes
+    IoReadsWrites = 0b10,
+
+    /// Data reads or writes but not instruction fetches
+    DataReadsWrites = 0b11,
+}
+
+impl HwBreakpointCondition {
+    const fn bit_range(n: DebugAddressRegisterNumber) -> Range<usize> {
+        let lsb = (16 + 4 * n.0) as usize;
+        lsb..lsb + 2
+    }
+}
+
+/// The size of a hardware breakpoint.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(u8)]
+pub enum HwBreakpointSize {
+    /// 1 byte length
+    Length1B = 0b00,
+
+    /// 2 byte length
+    Length2B = 0b01,
+
+    /// 8 byte length
+    Length8B = 0b10,
+
+    /// 4 byte length
+    Length4B = 0b11,
+}
+
+impl HwBreakpointSize {
+    /// Creates a new hardware breakpoint size if `size` is valid.
+    pub const fn new(size: usize) -> Option<Self> {
+        match size {
+            1 => Some(Self::Length1B),
+            2 => Some(Self::Length2B),
+            8 => Some(Self::Length8B),
+            4 => Some(Self::Length4B),
+            _ => None,
+        }
+    }
+
+    const fn bit_range(n: DebugAddressRegisterNumber) -> Range<usize> {
+        let lsb = (18 + 4 * n.0) as usize;
+        lsb..lsb + 2
+    }
+}
+
+impl TryFrom<usize> for HwBreakpointSize {
+    type Error = TryFromIntError;
+
+    fn try_from(size: usize) -> Result<Self, Self::Error> {
+        Self::new(size).ok_or(TryFromIntError(()))
+    }
+}
+
+/// A valid value of the [`Dr7`] debug register.
+///
+/// In addition to the [`Dr7Flags`] this value has a condition field and a size field for each debug address register.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Dr7Value {
+    bits: u64,
+}
+
+impl From<Dr7Flags> for Dr7Value {
+    fn from(dr7_flags: Dr7Flags) -> Self {
+        Self::from_bits_truncate(dr7_flags.bits())
+    }
+}
+
+impl Dr7Value {
+    const fn mask() -> u64 {
+        let field_mask = (1 << 32) - (1 << 16);
+        let flag_mask = Dr7Flags::all().bits();
+        field_mask | flag_mask
+    }
+
+    /// Convert from underlying bit representation, unless that representation contains bits that do not correspond to a field.
+    #[inline]
+    pub const fn from_bits(bits: u64) -> Option<Self> {
+        if (bits & Self::mask()) == 0 {
```

Comment:
Shouldn't this be

```suggestion
        if (bits & Self::mask()) == bits {
```

?

Reply:
Indeed! I changed it to `(bits & !Self::mask()) == 0` though, to match `bitflags`.